### ISSUE TITLE
IPv6 Unicast Interface

### ISSUE THREAD

**Issue by CDirkx** [rust-lang/rust#85604]
This issue is split out of the larger discussion around stabilization of the `ip` feature. The focus of this issue on the question of what interface Rust should provide for IPv6 unicast adresses.

The current unstable interface is as follows:

```rust
impl Ipv6Addr {
    fn is_unicast_global(&self) -> bool;
    fn is_unicast_link_local(&self) -> bool;
    fn is_unicast_link_local_strict(&self) -> bool;
    fn is_unicast_site_local(&self) -> bool;
}
```

Tracking issue for `ip`: #27709

### Behaviour of `is_unicast_link_local`/`is_unicast_link_local_strict`

Concern was raised about the need for both `is_unicast_link_local` and `is_unicast_link_local_strict`(https://github.com/rust-lang/rust/pull/66584#issuecomment-574036125). `is_unicast_link_local` tests if an address is in `FE80::/10`, `is_unicast_link_local_strict` in the stricter `FE80::/64`. For a time it was unclear which interpretation was the correct one, see https://github.com/rust-lang/rust/pull/76098#issue-475853748 for a more complete overview. However, it was mentioned (https://github.com/rust-lang/rust/pull/76098#issuecomment-705928605) that [IETF RFC #5156 Section 2.4](https://datatracker.ietf.org/doc/html/rfc5156#section-2.4) defines `FE80::/10` as the link-local unicast addresses, and that other programming languages and the linux kernel all use `FE80::/10` (https://github.com/rust-lang/rust/pull/76098#issuecomment-706916840, https://github.com/rust-lang/rust/pull/76098#issuecomment-705928605). The conclusion seems to be that the current implementation of `is_unicast_link_local` is correct and consistent with other implementations, `is_unicast_link_local_strict` could be removed.

Unresolved: Should `is_unicast_link_local_strict` be removed?

### Deprecation of site-local addresses

Unicast site-local addresses were deprecated in [IETF RFC #3879](https://datatracker.ietf.org/doc/html/rfc3879), see also [RFC #4291 Section 2.5.7](https://datatracker.ietf.org/doc/html/rfc4291#section-2.5.7). Any new implementation must no longer support the special behaviour of site-local addresses. This is mentioned in the docs of `is_unicast_site_local` and already implemented in `is_unicast_global`, which considers site-local addresses to be global.

For reference; .NET has [IsIPv6SiteLocal](https://docs.microsoft.com/en-us/dotnet/api/system.net.ipaddress.isipv6sitelocal?view=net-5.0), Python has [is_site_local](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv6Address.is_site_local) but mentions that is has been deprecated.

Unresolved: Should `is_unicast_site_local` be removed? Should `SiteLocal` be included in a possible `IPv6UnicastScope`?

### Introduce `IPv6UnicastScope`
It was suggested (https://github.com/rust-lang/rust/pull/76098#issuecomment-735019872) to replace the existing unicast interface with an enum `IPv6UnicastScope`, similar to `IPv6MulticastScope`:

```rust
impl Ipv6Addr {
    fn unicast_scope(&self) -> Option<IPv6UnicastScope>
}
```

Positive reaction (https://github.com/rust-lang/rust/pull/76098#issuecomment-735020358, https://github.com/rust-lang/rust/pull/76098#issuecomment-760762274)

Unresolved: What would be the definition of `IPv6UnicastScope`? 

## RFCs

- Unicast Adresses: [IETF RFC #4291 Section 2.5](https://datatracker.ietf.org/doc/html/rfc4291#section-2.5)
- Unique Local IPv6 Unicast Addresses: [IETF RFC #4193](https://datatracker.ietf.org/doc/html/rfc4193)
- Deprecating Site Local Addresses: [IETF RFC #3879](https://datatracker.ietf.org/doc/html/rfc3879)
- Special-Use IPv6 Addresses: [IETF RFC #5156](https://datatracker.ietf.org/doc/html/rfc5156)

## Previous Discussion

- stabilize the "ip" feature: [PR #66584](https://github.com/rust-lang/rust/pull/66584)
- Stabilize the "IP" feature: [PR #76098](https://github.com/rust-lang/rust/pull/76098)
- Stabilize the "IP" feature (continued from #76098): [PR #85585](https://github.com/rust-lang/rust/pull/85585)

**Comment by CDirkx** [comment]
With site-local unicast addresses deprecated, are the only two unicast scopes now link-local and global? If a unicast address is not-link-local, is it always global (and vice versa)?.

**Comment by CDirkx** [comment]
If I read the following table and the rest of RFC 4291 correctly:

```
Address type         Binary prefix        IPv6 notation   Section
------------         -------------        -------------   -------
Unspecified          00...0  (128 bits)   ::/128          2.5.2
Loopback             00...1  (128 bits)   ::1/128         2.5.3
Multicast            11111111             FF00::/8        2.7
Link-Local unicast   1111111010           FE80::/10       2.5.6
Global Unicast       (everything else)
```

1. Any address is either unicast or multicast
2. The unspecified and loopback address are unicast (They are defined under section 2.5 Unicast Addresses)
3. The unspecified and loopback address are not link-local unicast (link-local addresses have prefix `FE80::/10`), although of the loopback address it is stated that "it is treated as having Link-Local scope".
4. The unspecified and loopback address are not global unicast, as per the above table.

This leads to a problem if we would want to model unicast addresses similar to multicast addresses:

```rust 
enum Ipv6UnicastScope {
    LinkLocal,
    Global
}

impl Ipv6Addr {
    fn is_unicast(&self) -> bool;
    fn unicast_scope(&self) -> Option<Ipv6UnicastScope>
}
```

`Ipv6Addr::UNSPECIFIED.is_unicast()` would have to be `true` (2), but what would 
`Ipv6Addr::UNSPECIFIED.unicast_scope()` return? It can not return `Some(LinkLocal)` (3), and not `Some(Global)` (4). `None`?

**Comment by CDirkx** [comment]
My plan forward for the IPv6 unicast interface:

- Fix `is_unicast_global` to check for global unicast scope (currently it tries to check if the address appears globally reachable, which is not the same thing)
- Remove `is_unicast_link_local_strict`
- Remove `is_unicast_site_local`
- Add `is_unicast`

After those steps the interface will look like this:
```rust
impl Ipv6Addr {
    fn is_unicast(&self) -> bool;
    fn is_unicast_global(&self) -> bool;
    fn is_unicast_link_local(&self) -> bool;
}
```

and we can then consider replacing `is_unicast_global` and `is_unicast_link_local` with `unicast_scope` and 

```rust
#[non_exhaustive]
enum Ipv6UnicastScope {
    LinkLocal,
    Global
}
```
